File: problems/1136/solution.cpp
class Solution {
public:
    int minimumSemesters(int n, vector<vector<int>>& relations) {
        int ans = 0;
        vector<vector<int>> g(n);
        vector<int> indegree(n);
        queue<int> q;
        
        for (const vector<int> e: relations) {
            int u = e[0] - 1, v = e[1] - 1;
            
            g[u].push_back(v);
            indegree[v]++;
        }
        
        for (int i = 0; i < indegree.size(); i++) {
            if (not indegree[i]) {
                q.push(i);
            }
        }
        
        while (not q.empty()) {
            for (int i = q.size(); i > 0; i--) {
                for (const int& next: g[q.front()]) {
                    if (not --indegree[next]) {
                        q.push(next);
                    }
                }
                
                q.pop();
            }
            
            ans++;
        }
        
        if (any_of(begin(indegree), end(indegree), [](const int& node) {
            return node;
        })) {
            return -1;
        }
        
        return ans;
    }
};

File: problems/1136/solution.py
class Solution:
    def minimumSemesters(self, n: int, relations: List[List[int]]) -> int:
        ans, g, indegree = 0, [[] for _ in range(n)], [0] * n
        
        for [u, v] in relations:
            g[u - 1].append(v - 1)
            indegree[v - 1] += 1
        
        q = [i for i in range(n) if not indegree[i]]
        
        while q:
            nq = []
            ans += 1
            for node in q:
                for next in g[node]:
                    indegree[next] -= 1
                    
                    if not indegree[next]:
                        nq.append(next)
            
            q = nq
        
        if len([i for i in range(n) if indegree[i]]):
            return -1
        
        return ans